File: reqrio-java/src/main/java/org/xllgl2017/Url.java
package org.xllgl2017;

import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

import java.util.HashMap;

public class Url implements AutoCloseable {
    private Pointer raw;

    /// 初始化
    ///
    /// @param url :请求地址
    public Url(String url) throws Exception {
        PointerByReference err = new PointerByReference();
        this.raw = Session.INSTANCE.Url_new(url, err);
        util.check_err_pointer(err);
    }

    /// @param sni :域名
    public Url(String url, String sni) throws Exception {
        this(url);
        this.setSni(sni);
    }

    /// @param params url请求参数，value应为未编码
    public Url(String url_str, HashMap<String, String> params) throws Exception {
        this(url_str);
        for (String key : params.keySet()) {
            this.add_param(key, params.get(key));
        }
    }

    /// 添加一个请求参数，若这个值已存在则会被覆盖
    ///
    /// @param name  :参数名
    /// @param value :参数值
    public void add_param(String name, String value) throws Exception {
        if (this.raw == null) throw new Exception("Url had dropped");
        try {
            util.check_err(Session.INSTANCE.Url_add_param(this.raw, name, value));
        } catch (Exception e) {
            close();
            throw e;
        }
    }

    /// 删除一个参数
    ///
    /// @param name :待删除的参数名
    public void remove_param(String name) throws Exception {
        if (this.raw == null) throw new Exception("Url had dropped");
        try {
            util.check_err(Session.INSTANCE.Url_remove_param(this.raw, name));
        } catch (Exception e) {
            close();
            throw e;
        }
    }

    public Pointer getRaw() throws Exception {
        if (this.raw == null) throw new Exception("Url had dropped");
        return raw;
    }

    public void setRaw(Pointer raw) {
        this.raw = raw;
    }

    /// @param sni :域名，在使用ip url时设置
    public void setSni(String sni) throws Exception {
        if (this.raw == null) throw new Exception("Url had dropped");
        try {
            util.check_err(Session.INSTANCE.Url_set_sni(this.raw, sni));
        } catch (Exception e) {
            close();
            throw e;
        }
    }

    @Override
    public void close() {
        if (this.raw != null)
            Session.INSTANCE.Url_drop(this.raw);
        this.raw = null;
    }
}

File: reqrio-js/Url.js
const {library, check_error, ref_char_ptr} = require('./bindings')

class Url {

    /**初始化Url
     * @param {string} base_url
     * @param {object|null} params
     * @param {string|null} sni
     **/
    constructor(base_url, params = null, sni = null) {
        const errPtr = ref_char_ptr();
        this.ptr = library.Url_new(base_url, errPtr);
        check_error(errPtr.deref());
        if (params !== undefined && params !== null) {
            for (const [key, value] of Object.entries(params)) {
                this.add_param(key, JSON.stringify(value));
            }
        }
        if (sni !== undefined && sni !== null) {
            this.set_sni(sni)
        }
    }

    /**添加一个查询参数
     * @param {string} name
     * @param {string} value
     **/
    add_param(name, value) {
        const err = library.Url_add_param(this.ptr, name, value);
        check_error(err, this.close);
    }

    /**移除一个查询参数
     * @param {string} name
     **/
    remove_param(name) {
        const err = library.Url_remove_param(this.ptr, name);
        check_error(err, this.close);
    }

    /**为该url设置SNI，在使用ip地址URL时使用
     * @param {string} sni
     **/
    set_sni(sni) {
        const err = library.Url_set_sni(this.ptr, sni);
        check_error(err, this.close);
    }

    close() {
        if (this.ptr == null) return;
        library.Url_drop(this.ptr);
        this.ptr = null;
    }
}

module.exports = {Url}

